### PR TITLE
deployment: wait for VM to shut down

### PR DESCRIPTION
**Pull request by m-ildefons**
When issued a stop command, wait for the VM to shut down (up to one
minute) before continuing.

fixes: #500
Signed-off-by: Moritz Röhrich <moritz.rohrich@suse.com>

### CODE REVIEW

**Comment by kshtsk** [comment]
do you have any logs showing this patch fixes the #500 ?

File: seslib/deployment.py [hunk 1]
```diff
@@ -956,6 +956,8 @@ def _stop(self, node):
             Log.warning("Node '{}' is not running: current status '{}'"
                         .format(node, self.nodes[node].status))
             return
+        # Ugly hack to let ssh successfully exit before the connection is
+        # dropped during VM shutdown:
         ssh_cmd = self._ssh_cmd(node)
```
Comment (kshtsk) [inline seslib/deployment.py:961]: This is a really ugly hack.... maybe we can just drop it and replace with `poweroff` in combination with your status check fix?

**Comment by m-ildefons** [comment]
Sure, here is the log. Notice how `ses7p-test` fully starts up, because the race between shut-down and start-up has been eliminated:
```text
│ ~/repos/sesdev │► sesdev status
Found 10 deployments:

+--------------------+---------+---------+------------------------------------+----------------+-----------------+
| ID                 | Version | Status  | Nodes                              | Public Network | Cluster Network |
+--------------------+---------+---------+------------------------------------+----------------+-----------------+
| 24713-274668-ses6  | ses6    | stopped | master, node1, node2, node3        | 10.20.110.0/24 | 10.21.62.0/24   |
| 24713-274668-ses7  | ses7    | stopped | master, node1, node2, node3        | 10.20.162.0/24 | 10.21.133.0/24  |
| 25168-276288-ses7  | ses7    | stopped | master, node1, node2, node3        | 10.20.165.0/24 | 10.21.39.0/24   |
| 25207-276507-ses7  | ses7    | stopped | master, node1, node2, node3        | 10.20.190.0/24 | 10.21.50.0/24   |
| 25208-276511-ses7  | ses7    | stopped | master, node1, node2, node3        | 10.20.54.0/24  | 10.21.54.0/24   |
| 25209-276510-ses6  | ses6    | stopped | master, node1, node2, node3        | 10.20.92.0/24  | 10.21.3.0/24    |
| 25035-275941-ses7p | ses7p   | stopped | master, node1, node2, node3, node4 | 10.20.57.0/24  | 10.21.155.0/24  |
| 25036-275942-ses7p | ses7p   | stopped | master, node1, node2, node3        | 10.20.153.0/24 | 10.21.103.0/24  |
| bsc-1200064        | ses6    | stopped | master, node1, node2, node3        | 10.20.48.0/24  | 10.21.25.0/24   |
| ses7p-test         | ses7p   | running | master, node1, node2, node3        | 10.20.192.0/24 | 10.21.102.0/24  |
+--------------------+---------+---------+------------------------------------+----------------+-----------------+

│ ~/repos/sesdev │► sesdev stop ses7p-test ; sesdev start ses7p-test
Stopping 1 cluster
Stopping node master of deployment ses7p-test
Stopping node node1 of deployment ses7p-test
Stopping node node2 of deployment ses7p-test
Stopping node node3 of deployment ses7p-test
Deployment ses7p-test stopped!
Starting 1 cluster
=== Running shell command ===
vagrant up --no-destroy-on-error --no-provision
Bringing machine 'master' up with 'libvirt' provider...
Bringing machine 'node1' up with 'libvirt' provider...
Bringing machine 'node2' up with 'libvirt' provider...
Bringing machine 'node3' up with 'libvirt' provider...
==> node3: Starting domain.
==> master: Starting domain.
==> node3: Waiting for domain to get an IP address...
==> node1: Starting domain.
==> master: Waiting for domain to get an IP address...
==> master: Waiting for machine to boot. This may take a few minutes...
==> node2: Starting domain.
==> node1: Waiting for domain to get an IP address...
==> node2: Waiting for domain to get an IP address...
==> node2: Waiting for machine to boot. This may take a few minutes...
==> node1: Waiting for machine to boot. This may take a few minutes...
==> node3: Waiting for machine to boot. This may take a few minutes...
    master: SSH address: 192.168.121.86:22
    node1: SSH address: 192.168.121.143:22
    node3: SSH address: 192.168.121.33:22
    node1: SSH username: vagrant
    master: SSH username: vagrant
    node2: SSH address: 192.168.121.26:22
    node3: SSH username: vagrant
    node2: SSH username: vagrant
    node1: SSH auth method: private key
    node3: SSH auth method: private key
    master: SSH auth method: private key
    node2: SSH auth method: private key
    node2: Warning: Host unreachable. Retrying...
    node1: Warning: Host unreachable. Retrying...
    node1: Warning: Connection refused. Retrying...
    node2: Warning: Connection refused. Retrying...
    master: Warning: Connection refused. Retrying...
    node3: Warning: Connection refused. Retrying...
    node1: Warning: Connection refused. Retrying...
    node2: Warning: Connection refused. Retrying...
    node3: Warning: Connection refused. Retrying...
    master: Warning: Connection refused. Retrying...
==> node3: Machine booted and ready!
==> node3: Creating shared folders metadata...
==> master: Machine booted and ready!
==> master: Creating shared folders metadata...
==> node1: Machine booted and ready!
==> node1: Creating shared folders metadata...
==> node2: Machine booted and ready!
==> node2: Creating shared folders metadata...
==> node3: Machine not provisioned because `--no-provision` is specified.
==> node1: Machine not provisioned because `--no-provision` is specified.
==> node2: Machine not provisioned because `--no-provision` is specified.
==> master: Machine not provisioned because `--no-provision` is specified.
Deployment ses7p-test started!
│ ~/repos/sesdev │► sesdev status
Found 10 deployments:

+--------------------+---------+---------+------------------------------------+----------------+-----------------+
| ID                 | Version | Status  | Nodes                              | Public Network | Cluster Network |
+--------------------+---------+---------+------------------------------------+----------------+-----------------+
| 24713-274668-ses6  | ses6    | stopped | master, node1, node2, node3        | 10.20.110.0/24 | 10.21.62.0/24   |
| 24713-274668-ses7  | ses7    | stopped | master, node1, node2, node3        | 10.20.162.0/24 | 10.21.133.0/24  |
| 25168-276288-ses7  | ses7    | stopped | master, node1, node2, node3        | 10.20.165.0/24 | 10.21.39.0/24   |
| 25207-276507-ses7  | ses7    | stopped | master, node1, node2, node3        | 10.20.190.0/24 | 10.21.50.0/24   |
| 25208-276511-ses7  | ses7    | stopped | master, node1, node2, node3        | 10.20.54.0/24  | 10.21.54.0/24   |
| 25209-276510-ses6  | ses6    | stopped | master, node1, node2, node3        | 10.20.92.0/24  | 10.21.3.0/24    |
| 25035-275941-ses7p | ses7p   | stopped | master, node1, node2, node3, node4 | 10.20.57.0/24  | 10.21.155.0/24  |
| 25036-275942-ses7p | ses7p   | stopped | master, node1, node2, node3        | 10.20.153.0/24 | 10.21.103.0/24  |
| bsc-1200064        | ses6    | stopped | master, node1, node2, node3        | 10.20.48.0/24  | 10.21.25.0/24   |
| ses7p-test         | ses7p   | running | master, node1, node2, node3        | 10.20.192.0/24 | 10.21.102.0/24  |
+--------------------+---------+---------+------------------------------------+----------------+-----------------+
│ ~/repos/sesdev │► sesdev status ses7p-test

Current status of the 4 nodes of deployment "ses7p-test"

+--------+---------+----------------+-----------------+
| Node   | Status  | Public Address | Cluster Address |
+--------+---------+----------------+-----------------+
| master | running | 10.20.192.200  | None            |
| node1  | running | 10.20.192.201  | 10.21.102.201   |
| node2  | running | 10.20.192.202  | 10.21.102.202   |
| node3  | running | 10.20.192.203  | 10.21.102.203   |
+--------+---------+----------------+-----------------+
```

File: seslib/deployment.py (hunk 1 above)
Comment (m-ildefons) [reply seslib/deployment.py:961]: If I find some time, I might rewrite this section. But for now it works good enough. I've just put the comment there to explain future readers why it is done this way and not by simply issuing a `poweroff` via ssh.

**Comment by kshtsk** [comment]
Thanks, but I don't see in this log message `Node {node} successfully stopped.`

**Comment by m-ildefons** [comment]
It's an `info` level log message, which means by default it's not shown, you'd have to use `--verbose`:
```text
│ ~/repos/sesdev │► sesdev --verbose stop ses7p-test; sesdev --verbose start ses7p-test
INFO: Verbose mode: ON
INFO: Instantiating deployment ses7p-test
INFO: Running sync command in directory .: ['vagrant', 'box', 'list']
INFO: Running sync command in directory .: ['ssh-add', '-L']
INFO: found extra key cardno:000610664588
INFO: Running sync command in directory /home/moritz/.sesdev/ses7p-test: ['vagrant', 'status']
Stopping 1 cluster
Stopping node master of deployment ses7p-test
INFO: Running sync command in directory /home/moritz/.sesdev/ses7p-test: ['vagrant', 'ssh-config', 'master']
INFO: _ssh_cmd: ['ssh', 'root@192.168.121.81', '-i', '/home/moritz/.sesdev/ses7p-test/keys/sesdev', '-o', 'IdentitiesOnly yes', '-o', 'StrictHostKeyChecking no', '-o', 'UserKnownHostsFile /dev/null', '-o', 'PasswordAuthentication no', '-o', 'LogLevel ERROR']
INFO: Running sync command in directory .: ['ssh', 'root@192.168.121.81', '-i', '/home/moritz/.sesdev/ses7p-test/keys/sesdev', '-o', 'IdentitiesOnly yes', '-o', 'StrictHostKeyChecking no', '-o', 'UserKnownHostsFile /dev/null', '-o', 'PasswordAuthentication no', '-o', 'LogLevel ERROR', 'echo "sleep 2 && shutdown -h now" > /root/shutdown.sh && chmod +x /root/shutdown.sh']
INFO: Running sync command in directory /home/moritz/.sesdev/ses7p-test: ['vagrant', 'ssh-config', 'master']
INFO: _ssh_cmd: ['ssh', 'root@192.168.121.81', '-i', '/home/moritz/.sesdev/ses7p-test/keys/sesdev', '-o', 'IdentitiesOnly yes', '-o', 'StrictHostKeyChecking no', '-o', 'UserKnownHostsFile /dev/null', '-o', 'PasswordAuthentication no', '-o', 'LogLevel ERROR']
INFO: Running sync command in directory .: ['ssh', 'root@192.168.121.81', '-i', '/home/moritz/.sesdev/ses7p-test/keys/sesdev', '-o', 'IdentitiesOnly yes', '-o', 'StrictHostKeyChecking no', '-o', 'UserKnownHostsFile /dev/null', '-o', 'PasswordAuthentication no', '-o', 'LogLevel ERROR', 'nohup /root/shutdown.sh > /dev/null 2>&1 &']
INFO: Running sync command in directory /home/moritz/.sesdev/ses7p-test: ['vagrant', 'status']
INFO: Running sync command in directory /home/moritz/.sesdev/ses7p-test: ['vagrant', 'status']
INFO: Node master successfully stopped.
Stopping node node1 of deployment ses7p-test
INFO: Running sync command in directory /home/moritz/.sesdev/ses7p-test: ['vagrant', 'ssh-config', 'node1']
INFO: _ssh_cmd: ['ssh', 'root@192.168.121.50', '-i', '/home/moritz/.sesdev/ses7p-test/keys/sesdev', '-o', 'IdentitiesOnly yes', '-o', 'StrictHostKeyChecking no', '-o', 'UserKnownHostsFile /dev/null', '-o', 'PasswordAuthentication no', '-o', 'LogLevel ERROR']
INFO: Running sync command in directory .: ['ssh', 'root@192.168.121.50', '-i', '/home/moritz/.sesdev/ses7p-test/keys/sesdev', '-o', 'IdentitiesOnly yes', '-o', 'StrictHostKeyChecking no', '-o', 'UserKnownHostsFile /dev/null', '-o', 'PasswordAuthentication no', '-o', 'LogLevel ERROR', 'echo "sleep 2 && shutdown -h now" > /root/shutdown.sh && chmod +x /root/shutdown.sh']
INFO: Running sync command in directory /home/moritz/.sesdev/ses7p-test: ['vagrant', 'ssh-config', 'node1']
INFO: _ssh_cmd: ['ssh', 'root@192.168.121.50', '-i', '/home/moritz/.sesdev/ses7p-test/keys/sesdev', '-o', 'IdentitiesOnly yes', '-o', 'StrictHostKeyChecking no', '-o', 'UserKnownHostsFile /dev/null', '-o', 'PasswordAuthentication no', '-o', 'LogLevel ERROR']
INFO: Running sync command in directory .: ['ssh', 'root@192.168.121.50', '-i', '/home/moritz/.sesdev/ses7p-test/keys/sesdev', '-o', 'IdentitiesOnly yes', '-o', 'StrictHostKeyChecking no', '-o', 'UserKnownHostsFile /dev/null', '-o', 'PasswordAuthentication no', '-o', 'LogLevel ERROR', 'nohup /root/shutdown.sh > /dev/null 2>&1 &']
INFO: Running sync command in directory /home/moritz/.sesdev/ses7p-test: ['vagrant', 'status']
INFO: Running sync command in directory /home/moritz/.sesdev/ses7p-test: ['vagrant', 'status']
INFO: Running sync command in directory /home/moritz/.sesdev/ses7p-test: ['vagrant', 'status']
INFO: Node node1 successfully stopped.
Stopping node node2 of deployment ses7p-test
INFO: Running sync command in directory /home/moritz/.sesdev/ses7p-test: ['vagrant', 'ssh-config', 'node2']
INFO: _ssh_cmd: ['ssh', 'root@192.168.121.184', '-i', '/home/moritz/.sesdev/ses7p-test/keys/sesdev', '-o', 'IdentitiesOnly yes', '-o', 'StrictHostKeyChecking no', '-o', 'UserKnownHostsFile /dev/null', '-o', 'PasswordAuthentication no', '-o', 'LogLevel ERROR']
INFO: Running sync command in directory .: ['ssh', 'root@192.168.121.184', '-i', '/home/moritz/.sesdev/ses7p-test/keys/sesdev', '-o', 'IdentitiesOnly yes', '-o', 'StrictHostKeyChecking no', '-o', 'UserKnownHostsFile /dev/null', '-o', 'PasswordAuthentication no', '-o', 'LogLevel ERROR', 'echo "sleep 2 && shutdown -h now" > /root/shutdown.sh && chmod +x /root/shutdown.sh']
INFO: Running sync command in directory /home/moritz/.sesdev/ses7p-test: ['vagrant', 'ssh-config', 'node2']
INFO: _ssh_cmd: ['ssh', 'root@192.168.121.184', '-i', '/home/moritz/.sesdev/ses7p-test/keys/sesdev', '-o', 'IdentitiesOnly yes', '-o', 'StrictHostKeyChecking no', '-o', 'UserKnownHostsFile /dev/null', '-o', 'PasswordAuthentication no', '-o', 'LogLevel ERROR']
INFO: Running sync command in directory .: ['ssh', 'root@192.168.121.184', '-i', '/home/moritz/.sesdev/ses7p-test/keys/sesdev', '-o', 'IdentitiesOnly yes', '-o', 'StrictHostKeyChecking no', '-o', 'UserKnownHostsFile /dev/null', '-o', 'PasswordAuthentication no', '-o', 'LogLevel ERROR', 'nohup /root/shutdown.sh > /dev/null 2>&1 &']
INFO: Running sync command in directory /home/moritz/.sesdev/ses7p-test: ['vagrant', 'status']
INFO: Running sync command in directory /home/moritz/.sesdev/ses7p-test: ['vagrant', 'status']
INFO: Running sync command in directory /home/moritz/.sesdev/ses7p-test: ['vagrant', 'status']
INFO: Node node2 successfully stopped.
Stopping node node3 of deployment ses7p-test
INFO: Running sync command in directory /home/moritz/.sesdev/ses7p-test: ['vagrant', 'ssh-config', 'node3']
INFO: _ssh_cmd: ['ssh', 'root@192.168.121.164', '-i', '/home/moritz/.sesdev/ses7p-test/keys/sesdev', '-o', 'IdentitiesOnly yes', '-o', 'StrictHostKeyChecking no', '-o', 'UserKnownHostsFile /dev/null', '-o', 'PasswordAuthentication no', '-o', 'LogLevel ERROR']
INFO: Running sync command in directory .: ['ssh', 'root@192.168.121.164', '-i', '/home/moritz/.sesdev/ses7p-test/keys/sesdev', '-o', 'IdentitiesOnly yes', '-o', 'StrictHostKeyChecking no', '-o', 'UserKnownHostsFile /dev/null', '-o', 'PasswordAuthentication no', '-o', 'LogLevel ERROR', 'echo "sleep 2 && shutdown -h now" > /root/shutdown.sh && chmod +x /root/shutdown.sh']
INFO: Running sync command in directory /home/moritz/.sesdev/ses7p-test: ['vagrant', 'ssh-config', 'node3']
INFO: _ssh_cmd: ['ssh', 'root@192.168.121.164', '-i', '/home/moritz/.sesdev/ses7p-test/keys/sesdev', '-o', 'IdentitiesOnly yes', '-o', 'StrictHostKeyChecking no', '-o', 'UserKnownHostsFile /dev/null', '-o', 'PasswordAuthentication no', '-o', 'LogLevel ERROR']
INFO: Running sync command in directory .: ['ssh', 'root@192.168.121.164', '-i', '/home/moritz/.sesdev/ses7p-test/keys/sesdev', '-o', 'IdentitiesOnly yes', '-o', 'StrictHostKeyChecking no', '-o', 'UserKnownHostsFile /dev/null', '-o', 'PasswordAuthentication no', '-o', 'LogLevel ERROR', 'nohup /root/shutdown.sh > /dev/null 2>&1 &']
INFO: Running sync command in directory /home/moritz/.sesdev/ses7p-test: ['vagrant', 'status']
INFO: Running sync command in directory /home/moritz/.sesdev/ses7p-test: ['vagrant', 'status']
INFO: Running sync command in directory /home/moritz/.sesdev/ses7p-test: ['vagrant', 'status']
INFO: Running sync command in directory /home/moritz/.sesdev/ses7p-test: ['vagrant', 'status']
INFO: Node node3 successfully stopped.
Deployment ses7p-test stopped!
INFO: Verbose mode: ON
INFO: Instantiating deployment ses7p-test
INFO: Running sync command in directory .: ['vagrant', 'box', 'list']
INFO: Running sync command in directory .: ['ssh-add', '-L']
INFO: found extra key cardno:000610664588
INFO: Running sync command in directory /home/moritz/.sesdev/ses7p-test: ['vagrant', 'status']
Starting 1 cluster
INFO: Running async command in directory /home/moritz/.sesdev/ses7p-test: ['vagrant', 'up', '--no-destroy-on-error', '--no-provision']
=== Running shell command ===
vagrant up --no-destroy-on-error --no-provision
Bringing machine 'master' up with 'libvirt' provider...
Bringing machine 'node1' up with 'libvirt' provider...
Bringing machine 'node2' up with 'libvirt' provider...
Bringing machine 'node3' up with 'libvirt' provider...
==> node2: Starting domain.
==> node3: Starting domain.
==> master: Starting domain.
==> node3: Waiting for domain to get an IP address...
==> node2: Waiting for domain to get an IP address...
==> node1: Starting domain.
==> node3: Waiting for machine to boot. This may take a few minutes...
==> master: Waiting for domain to get an IP address...
==> master: Waiting for machine to boot. This may take a few minutes...
==> node1: Waiting for domain to get an IP address...
==> node1: Waiting for machine to boot. This may take a few minutes...
==> node2: Waiting for machine to boot. This may take a few minutes...
    node3: SSH address: 192.168.121.164:22
    master: SSH address: 192.168.121.81:22
    node2: SSH address: 192.168.121.184:22
    node3: SSH username: vagrant
    node1: SSH address: 192.168.121.50:22
    master: SSH username: vagrant
    master: SSH auth method: private key
    node2: SSH username: vagrant
    node1: SSH username: vagrant
    node1: SSH auth method: private key
    node3: SSH auth method: private key
    node2: SSH auth method: private key
    node2: Warning: Connection refused. Retrying...
    node3: Warning: Connection refused. Retrying...
    node1: Warning: Connection refused. Retrying...
    master: Warning: Connection refused. Retrying...
    node2: Warning: Connection refused. Retrying...
    master: Warning: Connection refused. Retrying...
    node1: Warning: Connection refused. Retrying...
    node3: Warning: Connection refused. Retrying...
==> node2: Machine booted and ready!
==> node2: Creating shared folders metadata...
==> node1: Machine booted and ready!
==> node3: Machine booted and ready!
==> node1: Creating shared folders metadata...
==> node3: Creating shared folders metadata...
==> master: Machine booted and ready!
==> master: Creating shared folders metadata...
==> node2: Machine not provisioned because `--no-provision` is specified.
==> node3: Machine not provisioned because `--no-provision` is specified.
==> node1: Machine not provisioned because `--no-provision` is specified.
==> master: Machine not provisioned because `--no-provision` is specified.
Deployment ses7p-test started!
```

**Comment by kshtsk** [comment]
Thanks Moritz, this is what I'd like to see.

**Comment by kshtsk** [comment]
jenkins retest this please

**Review by kshtsk — APPROVED** [review]

**Pull request merged**